src/serde_support.rs

```diff
@@ -0,0 +1,235 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+extern crate alloc;
+
+use alloc::format;
+use alloc::string::String;
+
+use dusk_bytes::Serializable;
+use serde::de::Error as SerdeError;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use crate::{
+    PublicKey, PublicKeyDouble, PublicKeyVarGen, SecretKey, SecretKeyVarGen,
+    Signature, SignatureDouble, SignatureVarGen,
+};
+
+impl Serialize for PublicKey {
+    fn serialize<S: Serializer>(
+        &self,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        let s = bs58::encode(self.to_bytes()).into_string();
+        serializer.serialize_str(&s)
+    }
+}
+
+impl<'de> Deserialize<'de> for PublicKey {
+    fn deserialize<D: Deserializer<'de>>(
+        deserializer: D,
+    ) -> Result<Self, D::Error> {
+        let s = String::deserialize(deserializer)?;
+        let decoded =
+            bs58::decode(&s).into_vec().map_err(SerdeError::custom)?;
+        let decoded_len = decoded.len();
+        let byte_length_str = format!("{}", Self::SIZE);
+        let bytes: [u8; Self::SIZE] = decoded.try_into().map_err(|_| {
+            SerdeError::invalid_length(decoded_len, &byte_length_str.as_str())
+        })?;
```

Comment:
Very sorry, but even though I suggested it, I'm afraid I changed my mind on the approach here.

The thing is that by using `try_into` to create a fixed size array from the vector, we won't error if `decoded.len() < Self::SIZE` even though we should.

```suggestion

        if decoded_len != Self::SIZE {
            let expected_str = format!("{}", Self::SIZE);
            return Err(SerdeError::invalid_length(
                decoded.len(),
                &expected_str.as_str(),
            ));
        }
        let mut bytes = [0u8; Self::SIZE];
        bytes.copy_from_slice(&decoded[..]);

```

I apologize for all the back and forth but since we will add `serde` support for a few types I want to make sure that this implementation is sound so that in the following PR's we can simply copy the same approach.

Reply:
It will error. Calling `try_into` to convert a vector to a fixed array only works if their length is exactly the same: [doc here](https://doc.rust-lang.org/std/primitive.array.html#impl-TryFrom%3CVec%3CT,+A%3E%3E-for-%5BT;+N%5D)

Reply:
Ah, nice! In that case I revert my reversal.